File: 1396-design-underground-system/1396-design-underground-system.java
class UndergroundSystem {
    HashMap<Integer,Pair<String,Integer>>hm;
    HashMap<String,ArrayList<Integer>>avg;
    public UndergroundSystem() {
        hm=new HashMap<>();
        avg=new HashMap<>();    
    }
    
    public void checkIn(int id, String stationName, int t) {
        Pair<String,Integer>Ckin= new Pair<String,Integer>(stationName,t); 
        hm.put(id,Ckin);
    }
    
    public void checkOut(int id, String stationName, int t) {
        Pair<String,Integer>Ckout=hm.get(id);
        StringBuilder path = new StringBuilder(Ckout.getKey());
        path.append("to:"+stationName);
        int timeT=t-Ckout.getValue();
        if(avg.containsKey(path.toString())){
            avg.get(path.toString()).add(timeT);
        }
        else{
            ArrayList<Integer>ar = new ArrayList<>();
            ar.add(t-Ckout.getValue());
            avg.put(path.toString(),ar);
        }
        hm.remove(id);
      
    }
    
    public double getAverageTime(String startStation, String endStation) {
        double sol=0;
        StringBuilder path = new StringBuilder(startStation);
        path.append("to:"+endStation);
        ArrayList<Integer>av = avg.get(path.toString());
        for(int i:av){
            sol+=i;
        }
        return sol/av.size();
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */

File: 1396-design-underground-system/1396-design-underground-system.cpp
#include <unordered_map>
#include <vector>
#include <string>

class UndergroundSystem {
    std::unordered_map<int, std::pair<std::string, int>> hm; // map id to start station and arrival time via pair
    std::unordered_map<std::string, std::vector<int>> avg; // map start station + "to:" + end station to an array of total time taken between routes
    
public:
    UndergroundSystem() {
        // Constructor
    }
    
    void checkIn(int id, std::string stationName, int t) {
        std::pair<std::string, int> Ckin(stationName, t);
        hm[id] = Ckin; // store the check-in details for the given id
    }
    
    void checkOut(int id, std::string stationName, int t) {
        std::pair<std::string, int> Ckout = hm[id]; // retrieve the check-in details for the given id
        std::string path = Ckout.first + "to:" + stationName; // create a path from start station to end station
        int timeT = t - Ckout.second; // calculate the total time taken between stations
        
        if (avg.find(path) != avg.end()) {
            avg[path].push_back(timeT); // if path exists, add the time taken to the existing list
        }
        else {
            std::vector<int> ar;
            ar.push_back(timeT); // create a new list and add the time taken
            avg[path] = ar;
        }
        
        hm.erase(id); // remove the id from the map as the journey is completed
    }
    
    double getAverageTime(std::string startStation, std::string endStation) {
        std::string path = startStation + "to:" + endStation;
        std::vector<int> av = avg[path]; // get the associated list of times for the path
        double sol = 0;
        
        for (int i : av) {
            sol += i; // calculate the sum of times
        }
        
        return sol / av.size(); // return the average time
    }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj;
 * obj.checkIn(id, stationName, t);
 * obj.checkOut(id, stationName, t);
 * double param_3 = obj.getAverageTime(startStation, endStation);
 */
